### PR TITLE
Basic virtual fn support in forward mode.

### PR DESCRIPTION
**Pull request by alexander-penev**
We can differentiate virtual class functions with several caveats:
```
struct A {
  virtual double f(double x) {}
  virtual double f_darg0(double x); // clad will provide a definition
};
```

### CODE REVIEW

File: demos/ComputerGraphics/SmallPT.cpp [hunk 1]
```diff
@@ -431,26 +387,26 @@ Vec radiance(const Ray &ray, int depth, unsigned short *Xi) {
...
+    double p = f.x>f.y && f.x>f.z ? f.x : f.y>f.z ? f.y : f.z; // max refl
 
     // object intersection id map test
     //return obj.c;
 
     cl = cl + cf.mult(obj.e);
 
-    if (++depth>5) {
-      if (erand48(Xi)<p) f=f*(1/p); else return cl;
+    if (++depth>5 || !p) {
```
Comment (vgvassilev) [inline demos/ComputerGraphics/SmallPT.cpp:397]: Why do we need this condition now?

File: demos/ComputerGraphics/SmallPT.cpp [hunk 2]
```diff
@@ -431,26 +387,26 @@ Vec radiance(const Ray &ray, int depth, unsigned short *Xi) {
...
+      if (erand48(Xi)<p) f=f*(1/p); else return cl*E;
```
Comment (vgvassilev) [inline demos/ComputerGraphics/SmallPT.cpp:398]: Is that change intended?

**Review by vgvassilev — COMMENTED** [review]
This is really great! Thanks for working on this!!

File: demos/ComputerGraphics/SmallPT.cpp (hunk 2 above)
Comment (alexander-penev) [reply demos/ComputerGraphics/SmallPT.cpp:398]: This in not needed anymore. Removed.

File: demos/ComputerGraphics/SmallPT.cpp (hunk 1 above)
Comment (alexander-penev) [reply demos/ComputerGraphics/SmallPT.cpp:397]: Not needed. Source is changed now.

**Review by vgvassilev — APPROVED** [review]
LGTM! Thanks!
